userspace/src/lib.rs

```diff
@@ -0,0 +1,37 @@
+#![no_std]
+#![feature(lang_items)]
+#![feature(start)]
+#![feature(wasm_import_module)]
+
+#[lang = "panic_fmt"]
+fn panic_fmt() -> ! {
+    loop {}
+}
+
+#[lang = "oom"]
+fn oom() -> ! {
+    loop {}
+}
+
+// This does nothing but satisfy the rust compiler
+// (well actually it exports a `main` function that calls this, but it doesn't
+// have the type we want it to have).
+#[start]
+pub fn rust_start_not_called(_argc: isize, _argv: *const *const u8) -> isize {
+    0xbadbad
+}
+
+pub mod abi {
+    #[wasm_import_module = "abi"]
+    extern {
+        pub fn print(ptr: i32, len: i32);
```

Comment:
The signature for this can be `pub fn print(ptr: *mut u8, len: usize)`.

Reply:
The `*mut` part definitely seems like the right decision. Will update.

I'm not sure that representing `len` as a unsigned type makes sense, when it's treated as signed in the kernel code.

Reply:
Where is it treated as signed? IMO length should definitely be unsigned, since a negative length is obviously invalid. Since it's being interpreted as unsigned in the kernel, it'll just wrap.

Reply:
Ah, wait, you're right. I'm completely wrong about code I just wrote.

I wrote this code after looking at [abi_map!](https://github.com/gmorenz/nebulet/blob/95c07e454bf793d7746dd2afbb78d850369918c9/src/wasm/abi.rs#L52) and seeing `I32` (which I had copied from other abi things using a buffer) and naively assumed that that meant my rust function was taking an `i32`, even though I wrote the rust function to take a `u32` (also copying other things taking a buffer).

Will change.